rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,484 @@
+
+# Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+## Interaction with other features
+The e2e conformance tests proposed would be carried out outside the Linkerd CI. These tests would make use of the `kubectl` and `linkerd` binaries to interact with the cluster and with various features Linkerd can perform. The testing process can be configured easily using a YAML definition file (more details explained under “implementation details”). Such an approach gives the users the flexibility in diagnosing the state of their cluster w.r.t. Linkerd. 
+
+We initially start off by modifying the existing emojivoto application to have feature flags to enable/disable features such as MySQL, Redis, gRPC, websockets, etc. This is important because emojivoto is heavily used in the getting started process.
+
+## Implementation Details
+The test suite and its dependencies shall be packaged as a Docker container. This container shall run as a job in the cluster. In the Dockerfile, we define that the following must happen :
+
+- Setup and install all the binaries/tools required by the testing environment - `linkerd`, `kubectl`, `mysql`, etc.
+- Copy the test configuration file _conformance_config.yaml_ which is responsible for describing how the tests must run (see below for implementation details).
+- Copy _run.sh_ file that initiates the tests using the `go test` cmd.
+- Copy the test suite
+- ENTRYPOINT / CMD that executes `./run.sh`
+
+Users may easily run this container as a job using [Sonobuoy](https://sonobuoy.io/). It is a great tool when it comes to conformance validation, and its [plugin model](https://sonobuoy.io/docs/master/plugins/) can allow us to easily extend functionality beyond k8s conformance validtion. Sonobuoy also saves developers the overhead of configuring the desired objects such as Namespaces, ServieAccounts, ClusterRoleBindings, Jobs, RBAC, etc.
+
+> Note: When generating a Sonobuoy plugin, the plugin definition file sets the start-up command as ./run.sh by default. If we set the container’s ENTRYPOINT/CMD as `go test`, this may get overridden to `./run.sh` because of how Sonobuoy generates plugins. The generated plugin definition file may however be modified to not override the startup command of the container, but this may have a negative impact on the user experience.
+
+In order to run the Sonobuoy plugin, users/testers shall be required to carry out the following steps:
+```bash
+# setup docker registry
+$ export USER=<your public registry>
+$ docker build . -t $USER/progress-reporter:v0.1
```

Comment:
Hmm, Asking users to build the image just to pass some configuration, feels a bit hard to me. Can we use something like configmap attaching as volume, etc to make the container read the config during runtime? Yes I see that the plugin yaml is being generated later in the pipeline. Can we use kustomize to override the generated yaml with attaching a configmap?

If we don't find anything to replace, we can go with this way I guess!

Reply:
As per @ihcsim 's suggestion, I have removed this part and instead made the `.run.sh` file accept a list of arguments. Could you take a look? :smile: